### PR TITLE
[mlir][sparse] use a consistent order between [dis]assembleOp and sto…

### PR DESCRIPTION
**Pull request by PeimingLiu**
…rage layout.

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp [hunk 1]
```diff
@@ -33,20 +33,14 @@ static void convTypes(TypeRange types, SmallVectorImpl<Type> &convTypes,
...
-
     // Convert the external representation of the position/coordinate array.
     foreachFieldAndTypeInSparseTensor(stt, [&convTypes, extraTypes](
                                                Type t, FieldIndex,
                                                SparseTensorFieldKind kind,
                                                Level, LevelType) {
       if (kind == SparseTensorFieldKind::CrdMemRef ||
-          kind == SparseTensorFieldKind::PosMemRef) {
+          kind == SparseTensorFieldKind::PosMemRef ||
+          kind == SparseTensorFieldKind::ValMemRef) {
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp:43]: Oh, this is nice! Even less code.

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp [hunk 2]
```diff
@@ -33,20 +33,14 @@ static void convTypes(TypeRange types, SmallVectorImpl<Type> &convTypes,
     }
     // Convert the external representation of the values array.
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp:34]: This comment can be removed now

maybe move the single remaining sst L35 after comment at L36 to get a clean new section

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp [hunk 3]
```diff
@@ -73,34 +67,27 @@ static void convVals(OpBuilder &builder, Location loc, TypeRange types,
...
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp:67]: Change this comment to

Handle sparse data.

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp [hunk 4]
```diff
@@ -73,34 +67,27 @@ static void convVals(OpBuilder &builder, Location loc, TypeRange types,
...
+          kind == SparseTensorFieldKind::ValMemRef) {
         if (isIn) {
           inputs.push_back(fromVals[idx++]);
         } else {
           ShapedType st = t.cast<ShapedType>();
           auto rtp = RankedTensorType::get(st.getShape(), st.getElementType());
           inputs.push_back(extraVals[extra++]);
           retTypes.push_back(rtp);
-          cntTypes.push_back(rtp.getElementType());
+          cntTypes.push_back(builder.getIndexType());
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseAssembler.cpp:90]: is this correct? is it always an index type or did we use the same type as metadata?

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorCodegen.cpp [hunk 5]
```diff
@@ -1409,14 +1409,10 @@ struct SparseDisassembleOpConverter
...
-        // operand in unpack operation.
-        // TODO: maybe change unpack/pack operation instead to be
-        // consistent.
-        retMem.insert(retMem.begin(), dst);
+
+        retMem.push_back(dst);
         Type valLenTp = op.getValLen().getType();
-        retLen.insert(retLen.begin(),
-                      genScalarToTensor(rewriter, loc, sz, valLenTp));
+        retLen.push_back(genScalarToTensor(rewriter, loc, sz, valLenTp));
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorCodegen.cpp:1415]: yeah! I never liked this expensive insert at index 0 ;-)

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp [hunk 6]
```diff
@@ -738,12 +738,6 @@ class SparseTensorDisassembleConverter
     auto stt = getSparseTensorType(op.getTensor());
     SmallVector<Value> retVal;
     SmallVector<Value> retLen;
-    // Get the values buffer first.
-    auto vals = genValuesCall(rewriter, loc, stt, adaptor.getTensor());
-    auto valLenTp = op.getValLen().getType();
-    auto valLen = linalg::createOrFoldDimOp(rewriter, loc, vals, 0);
-    retVal.push_back(vals);
-    retLen.push_back(genScalarToTensor(rewriter, loc, valLen, valLenTp));
     // Then get the positions and coordinates buffers.
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp:741]: Remove the "Then" now that we don't have the first Get

**Review by aartbik — APPROVED** [review]

**Review by yinying-lisa-li — APPROVED** [review]

**Pull request merged**